File: 0653-two-sum-iv-input-is-a-bst/0653-two-sum-iv-input-is-a-bst.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {boolean}
 */
var findTarget = function(root, k) {
    const set = new Set()
    return dfs(root)
    
    function dfs(root) {
        if (!root) return false
        if (set.has(k - root.val)) return true
        set.add(root.val)
        return dfs(root.left) || dfs(root.right)
    }
};

findTarget = function(root, k) {
    const inOrdered = inOrder(root)
    let l = 0, r = inOrdered.length-1
    while (l < r) {
        if (inOrdered[l] + inOrdered[r] === k) {
            return true
        } else if (inOrdered[l] + inOrdered[r] > k) {
            r--
        } else {
            l++
        }
    }
    return false
    
    function inOrder(root) {
        if (!root) return []
        return [...inOrder(root.left), root.val, ...inOrder(root.right)]
    }
}

File: 0653-two-sum-iv-input-is-a-bst/0653-two-sum-iv-input-is-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findTarget(self, root: Optional[TreeNode], k: int) -> bool:
        seen = set()
        
        def dfs(root: Optional[TreeNode]) -> bool:
            nonlocal seen
            if not root:
                return False
            if k - root.val in seen:
                return True
            seen.add(root.val)
            return dfs(root.left) or dfs(root.right)
        
        return dfs(root)